### PR TITLE
Traffic mirroring

### PR DESCRIPTION
**Pull request by jaysheth2**
Initial release of Traffic-Mirroring

### CODE REVIEW

File: traffic-mirroring/mirroring.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+static char map_file[MAP_PATH_MAX];
+static bool chain = false;
+static int verbose = 1;
+
+static int tc_attach_bpf(char* dev, char* bpf_obj, char* dir);
+static bool validate_ifname(const char* input_ifname, char *output_ifname);
+static bool validate_address(char* input_address, network_addr_t output_address[], int *count);
+// static void update_ports(int map_fd, char *ports, char *prog_type, char *port_type);
+// static void bin(unsigned n);
+// static void printKthBit(unsigned int n, unsigned int k)
```
Comment (jniesz) [inline traffic-mirroring/mirroring.h:77]: this code needs to get removed.

**Review by jniesz — APPROVED** [review]

File: traffic-mirroring/Makefile [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+TOOLS_PATH := $(BPF_SAMPLES_PATH)/../../tools
+L3AF_SRC_PATH := $(CURDIR)
+
+# List of programs to build
+hostprogs-y += mirroring
+
+# Libbpf dependencies
+LIBBPF = $(TOOLS_PATH)/lib/bpf/libbpf.a
+
+mirroring-objs := mirroring.o ../bpf_load.o
```
Comment (sanfern) [inline traffic-mirroring/Makefile:15]: Can we change the path ../bpf_load.o to ../../bpf_load.o

**Review by jniesz — APPROVED** [review]

**Review by sanfern — APPROVED** [review]
LGTM

File: traffic-mirroring/mirroring.c [hunk 3]
```diff
@@ -0,0 +1,1226 @@
...
+{
+    char *ptr, *tmp;
+    uint32_t proto_num = 0;
+    uint32_t pval = 1;
+    tmp = strdup(protocols);
+    int temp = -555;
+    while ((ptr = strsep(&tmp, delim)) != NULL) {
+        ptr = trim_space(ptr);
+        if (strcmp(ptr, "udp") == 0) {
+            proto_num = (uint16_t)(strtoi("17"));
```
Comment (dthaler) [inline traffic-mirroring/mirroring.c:764]: can't you just use a common define like IPPROTO_UDP?

File: traffic-mirroring/mirroring.c [hunk 4]
```diff
@@ -0,0 +1,1226 @@
...
+            ret = EXIT_FAILURE;
+            exit_clean(ret);
+        }
+        if (verbose) {
+            log_info("Change egress redirect ifindex to: %d\n", tunnel_ifindex);
+        }
+    }
+
+    if (src_addr_filter) {
+        log_info("Source addres filter is set\n");
```
Comment (dthaler) [inline traffic-mirroring/mirroring.c:1084]: typo: "addres"

File: traffic-mirroring/mirroring.c [hunk 5]
```diff
@@ -0,0 +1,1226 @@
...
+                    free(src_key);
+                    exit_clean(ret);
+                }
+            }
+            free(src_key);
+        }
+    }
+
+    if (dst_addr_filter) {
+        log_info("Destination addres filter is set\n");
```
Comment (dthaler) [inline traffic-mirroring/mirroring.c:1128]: typo "addres"

**Review by dthaler — COMMENTED** [review]
Looks like this was already merged, but this PR has bugs and typos that need to be fixed.

**Pull request merged**